include/wil/network.h

```diff
@@ -0,0 +1,1328 @@
+        return ::wil::network::resolve_address(local_address_name_string);
+    }
+
+    // wil function to capture resolving the local-host (loopback) addresses, throwing on error
+    // returning an RAII object containing the results
+    inline ::wil::unique_addrinfo resolve_localhost_addresses()
+    {
+        constexpr auto* localhost_address_name_string = L"localhost";
+        return ::wil::network::resolve_address(localhost_address_name_string);
+    }
+#endif
+
+    struct WINSOCK_EXTENSION_FUNCTION_TABLE
+    {
+        LPFN_ACCEPTEX AcceptEx{nullptr};
+        LPFN_CONNECTEX ConnectEx{nullptr};
+        LPFN_DISCONNECTEX DisconnectEx{nullptr};
+        LPFN_GETACCEPTEXSOCKADDRS GetAcceptExSockaddrs{nullptr};
+        LPFN_TRANSMITFILE TransmitFile{nullptr};
+        LPFN_TRANSMITPACKETS TransmitPackets{nullptr};
+        LPFN_WSARECVMSG WSARecvMsg{nullptr};
+        LPFN_WSASENDMSG WSASendMsg{nullptr};
+    };
+
+    template <typename F>
+    struct socket_extension_function_table_t;
+    using winsock_extension_function_table = socket_extension_function_table_t<WINSOCK_EXTENSION_FUNCTION_TABLE>;
+    using rio_extension_function_table = socket_extension_function_table_t<RIO_EXTENSION_FUNCTION_TABLE>;
+
+    template <typename F>
+    struct socket_extension_function_table_t
+    {
+        static socket_extension_function_table_t load() WI_NOEXCEPT;
+
+        // can copy, but the new object needs its own WSA reference count
+        // (getting a WSA reference count should be no-fail once the first reference it taken)
+        //
+        // IF the target could not get a WSA reference count
+        //   OR
+        // IF we couldn't get our own WSA reference count
+        //   (this should never happen the caller has a reference, but we failed to get a WSA reference)
+        // THEN
+        //   this object cannot carry forward any function pointers - it must show successfully loaded == false
+        socket_extension_function_table_t(const socket_extension_function_table_t& rhs) WI_NOEXCEPT :
+            wsa_reference_count{WSAStartup_nothrow()}
+        {
+            if (!wsa_reference_count || !rhs.wsa_reference_count)
+            {
+                ::memset(&function_table, 0, sizeof(function_table));
+            }
+            else
+            {
+                ::memcpy_s(&function_table, sizeof(function_table), &rhs.function_table, sizeof(rhs.function_table));
+            }
+        }
+
+        socket_extension_function_table_t& operator=(const socket_extension_function_table_t& rhs) WI_NOEXCEPT
+        {
+            if (!wsa_reference_count || !rhs.wsa_reference_count)
+            {
+                ::memset(&function_table, 0, sizeof(function_table));
+            }
+            else
+            {
+                ::memcpy_s(&function_table, sizeof(function_table), &rhs.function_table, sizeof(rhs.function_table));
+            }
+
+            return *this;
+        }

```

Comment:
Consider:

Add move constructor/assignment operator that also takes ownership of the `unique_wsacleanup_call`. This would mostly just buy you one less `WSAStartup` call (since the function pointers are copied), so not essential, but nice to have.

NOTE: You can _technically_ define these as `= default`, and they will _technically_ do the correct thing, however the moved-from type will (potentially) still hold non-null pointers in its table. Nobody should be using a moved-from type, so it _shouldn't_ matter, but if they do, might result in tricky to debug issues.

Reply:
wil::unique_call really does not want its move c'tor called :(

so far I cannot get this to compile using the move c'tors for wil::unique_call. e.g., even this fails to compile

        socket_extension_function_table_t() WI_NOEXCEPT :
            wsa_reference_count{WSAStartup_nothrow()}
        {
            load();
            const auto test = ::wistd::move(wsa_reference_count); // I added this to just test to see if I could force the operator=(unique_call&&) --> but even this fails to compile ... I don't know what I'm missing :(
        }